Replace 'scale' with two independent parameters: 'iscale' and 'kscale'

mcara opened the pull request:
This PR deprecates parameter `scale` of `Drizzle.add_image()` method. This parameter has the meaning of pixel scale ratio and it was carried over from `drizzlepac` where it is used to do two separate things:
- to rescale input image by `scale**2` because there `PHOTFLAM` of the resampled image -- conversion factor used to convert count rate to flux units -- was not rescaled (and left equal to that of the first input image);
- to rescale resample kernels to the correct size.

These are two different and independent issues and having a single `scale` parameter is just incorrect for a wide range of cases. For example, in JWST, pixel values are surface brightness and converting to flux is achieved by multiplying them by pixel area which is modified for the output image (basically `PHOTFLAM` equivalent is modified) and so a `scale=1` is what is used there. However, this leads to incorrect kernel size, see, e.g., https://github.com/spacetelescope/stcal/pull/412

Splitting `scale` into two independent parameters will allow us to correctly handle various situations/needs.

mcara committed:
Replace 'scale' with two independent parameters: 'iscale' and 'kscale'

mcara committed:
fix some compiler warnings

braingram commented on drizzle/resample.py:
```diff
@@ -563,9 +565,10 @@ def _increment_ctx_id(self):
 
-    def add_image(self, data, exptime, pixmap, data2=None, scale=1.0,
+    def add_image(self, data, exptime, pixmap, data2=None,
+                  iscale=1.0, kscale=1.0,
```
```suggestion
                  iscale=1.0, kscale=1.0,
```
Maybe `input_scale_factor` and `kernel_scale_factor`?

perrygreenfield commented:
As for using pixfrac for kscale, I would argue against that since it would be completely non-obvious that it means something else in a different context. And pixfrac already has a long usage as a name. I would back up keeping these separate parameters and having the software raise an error if both are specified out of context.

perrygreenfield reviewed (approved):
LGTM

mcara commented:
> Mostly I'm trying to understand at the moment if they are in fact mutually exclusive parameters that control related concepts.

Although at first glance these parameters (`kscale` and `pixfrac`) seem to be related -- `pixfrac` controls pixel size of the single pixel in a kernel like "square" and "turbo" while `kscale` controls kernel size for larger (than one pixel) kernels like Lanczos -- I think conceptually they are distinct. For example, in principle, we could get rid of `kscale` parameter because it is used to convert "distances" (or "sizes") from the output image to the input image by using inverse transformation of the `pixmap` - then we would not even need this scale. However, even then, `pixfrac` would still make sense and be used in the square and turbo kernels.

We could also rename `iscale` to `"input_image_pre_resample_multiplicative_factor"` (or drop some of the words, like `pre_factor` but I don't think it would be clearer) and `kscale` - 'pixel_scale_ratio` (that's easier).

mcara committed:
Fix rebase mistakes

mcara committed:
Change parameter names. Deprecate and simplify tblot parameters

mcara commented:
@braingram @perrygreenfield 

I have updated documentation -- in particular see https://github.com/spacetelescope/drizzle/blob/c5e9095de4e36edc13bdd9910f489e6976dbbdae/drizzle/resample.py#L60-L99 -- and deprecated several parameters in `tblot` that are not really needed and renamed `kscale` as `pixel_scale_ratio`. Please review these lates changes. Thank you!

mcara committed:
Remove unnecessary loop for unit conversion in tdriz

braingram commented on drizzle/resample.py:
```diff
@@ -889,8 +980,9 @@ def add_image(self, data, exptime, pixmap, data2=None, dq=None, scale=1.0,
 
-def blot_image(data, pixmap, pix_ratio, exptime, output_pixel_shape,
-               interp='poly5', sinscl=1.0):
+def blot_image(data, pixmap, pix_ratio=_DEPRECATED_ARG,
```
For context, is this used anywhere outside drizzle?

braingram reviewed (commented):
Thanks. The docs updates look great and are very clear at explaining when iscale is useful.

Would you run romancal and jwst regtests with this PR?

mcara committed:
Further changes to tblot and blot_image

mcara committed:
Update docstring

mcara commented:
[Romancal regression tests pass](https://github.com/spacetelescope/RegressionTests/actions/runs/18666992170/job/53222048870) if https://github.com/spacetelescope/romancal/pull/2009 and https://github.com/spacetelescope/stcal/pull/418 are merged.

mcara committed:
attempt 1 to fix failing macosx test

mcara committed:
Fix some bugs revealed in testing
